File: results/2026-04-17_004319/tasks/11-semantic-version-bumper/typescript-bun-sonnet46-1m-medium/generated-code/run-act-tests.ts
// Act test harness — runs each test fixture through the GitHub Actions workflow via act
// Asserts on exact expected output values. Saves all output to act-result.txt.

import { spawnSync } from "child_process";
import { writeFileSync, readFileSync, mkdirSync, rmSync, existsSync, copyFileSync } from "fs";
import { join } from "path";

const ROOT = join(import.meta.dir);
const ACT_RESULT_FILE = join(ROOT, "act-result.txt");

// Clear previous results
writeFileSync(ACT_RESULT_FILE, "");

interface TestCase {
  name: string;
  initialVersion: string;
  gitLog: string;
  expectedNewVersion: string;
  expectedBumpType: string;
}

const testCases: TestCase[] = [
  {
    name: "patch-bump-fix-commits",
    initialVersion: "1.2.3",
    gitLog: [
      "abc1234 fix: resolve null pointer in parser",
      "abc1235 fix: handle empty string input",
    ].join("\n"),
    expectedNewVersion: "1.2.4",
    expectedBumpType: "patch",
  },
  {
    name: "minor-bump-feat-commits",
    initialVersion: "1.1.0",
    gitLog: [
      "def1234 feat: add support for pre-release versions",
      "def1235 fix: correct semver comparison",
    ].join("\n"),
    expectedNewVersion: "1.2.0",
    expectedBumpType: "minor",
  },
  {
    name: "major-bump-breaking-change",
    initialVersion: "2.3.1",
    gitLog: [
      "ghi1234 feat!: redesign API surface",
      "ghi1235 feat: add new flags",
    ].join("\n"),
    expectedNewVersion: "3.0.0",
    expectedBumpType: "major",
  },
];

// Static workflow that reads commits from commits.txt in the repo
const TEST_WORKFLOW = `name: Semantic Version Bumper

on:
  push:
    branches:
      - main
      - master
  workflow_dispatch:

permissions:
  contents: write

jobs:
  test:
    name: Run Tests
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Bun
        uses: oven-sh/setup-bun@v2
        with:
          bun-version: latest

      - name: Install dependencies
        run: bun install

      - name: Run unit tests
        run: bun test src/version-bumper.test.ts

  bump-version:
    name: Bump Version
    runs-on: ubuntu-latest
    needs: test
    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Bun
        uses: oven-sh/setup-bun@v2
        with:
          bun-version: latest

      - name: Install dependencies
        run: bun install

      - name: Run semantic version bump
        id: bump
        run: |
          OUTPUT=$(bun run src/main.ts version.txt commits.txt)
          echo "$OUTPUT"
          NEW_VERSION=$(echo "$OUTPUT" | grep "^New version:" | awk '{print $NF}')
          BUMP_TYPE=$(echo "$OUTPUT" | grep "^Bump type:" | awk '{print $NF}')
          echo "new_version=\${NEW_VERSION}" >> "$GITHUB_OUTPUT"
          echo "bump_type=\${BUMP_TYPE}" >> "$GITHUB_OUTPUT"

      - name: Print new version
        run: |
          echo "Version bump complete"
          echo "New version: \${{ steps.bump.outputs.new_version }}"
          echo "Bump type: \${{ steps.bump.outputs.bump_type }}"
`;

function appendToResults(content: string) {
  const existing = readFileSync(ACT_RESULT_FILE, "utf8");
  writeFileSync(ACT_RESULT_FILE, existing + content);
}

function runActTestCase(tc: TestCase): { success: boolean; output: string } {
  const tmpDir = join(ROOT, `tmp-act-${tc.name}`);

  try {
    mkdirSync(tmpDir, { recursive: true });
    mkdirSync(join(tmpDir, "src"), { recursive: true });
    mkdirSync(join(tmpDir, ".github/workflows"), { recursive: true });

    // Copy source files
    const filesToCopy = [
      "src/version-bumper.ts",
      "src/version-bumper.test.ts",
      "src/fixtures.ts",
      "src/types.ts",
      "src/main.ts",
      "package.json",
      "tsconfig.json",
    ];

    for (const f of filesToCopy) {
      const src = join(ROOT, f);
      const dst = join(tmpDir, f);
      if (existsSync(src)) {
        copyFileSync(src, dst);
      }
    }

    if (existsSync(join(ROOT, "bun.lockb"))) {
      copyFileSync(join(ROOT, "bun.lockb"), join(tmpDir, "bun.lockb"));
    }

    // Write fixture data for this test case
    writeFileSync(join(tmpDir, "version.txt"), tc.initialVersion + "\n");
    writeFileSync(join(tmpDir, "commits.txt"), tc.gitLog + "\n");

    // Write the static workflow (no dynamic YAML generation)
    writeFileSync(join(tmpDir, ".github/workflows/semantic-version-bumper.yml"), TEST_WORKFLOW);

    // Copy .actrc
    copyFileSync(join(ROOT, ".actrc"), join(tmpDir, ".actrc"));

    // Initialize git repo
    spawnSync("git", ["init"], { cwd: tmpDir });
    spawnSync("git", ["config", "user.email", "test@test.com"], { cwd: tmpDir });
    spawnSync("git", ["config", "user.name", "Test"], { cwd: tmpDir });
    spawnSync("git", ["add", "."], { cwd: tmpDir });
    spawnSync("git", ["commit", "-m", "chore: initial commit"], { cwd: tmpDir });

    // Run act
    const actResult = spawnSync(
      "act",
      ["push", "--rm", "--no-cache-server", "--pull=false"],
      {
        cwd: tmpDir,
        encoding: "utf8",
        timeout: 300000,
      }
    );

    const output = (actResult.stdout || "") + (actResult.stderr || "");

    return {
      success: actResult.status === 0,
      output,
    };
  } finally {
    rmSync(tmpDir, { recursive: true, force: true });
  }
}

let allPassed = true;

for (const tc of testCases) {
  console.log(`\n${"=".repeat(60)}`);
  console.log(`Running test case: ${tc.name}`);
  console.log(`  Initial version: ${tc.initialVersion}`);
  console.log(`  Expected new version: ${tc.expectedNewVersion}`);
  console.log(`  Expected bump type: ${tc.expectedBumpType}`);

  const delimiter = `\n${"=".repeat(60)}\nTEST CASE: ${tc.name}\nInitial: ${tc.initialVersion} -> Expected: ${tc.expectedNewVersion} (${tc.expectedBumpType})\n${"=".repeat(60)}\n`;
  appendToResults(delimiter);

  const { success, output } = runActTestCase(tc);
  appendToResults(output);

  if (!success) {
    console.error(`  FAIL: act exited non-zero for ${tc.name}`);
    appendToResults(`\nFAIL: act exited non-zero\n`);
    allPassed = false;
    continue;
  }

  if (!output.includes("Job succeeded")) {
    console.error(`  FAIL: 'Job succeeded' not found in act output for ${tc.name}`);
    appendToResults(`\nFAIL: 'Job succeeded' not found\n`);
    allPassed = false;
    continue;
  }

  if (!output.includes(tc.expectedNewVersion)) {
    console.error(`  FAIL: expected version ${tc.expectedNewVersion} not found in output`);
    appendToResults(`\nFAIL: expected version ${tc.expectedNewVersion} not found\n`);
    allPassed = false;
    continue;
  }

  if (!output.includes(tc.expectedBumpType)) {
    console.error(`  FAIL: expected bump type ${tc.expectedBumpType} not found in output`);
    appendToResults(`\nFAIL: expected bump type ${tc.expectedBumpType} not found\n`);
    allPassed = false;
    continue;
  }

  console.log(`  PASS: version=${tc.expectedNewVersion}, bump=${tc.expectedBumpType}, Job succeeded`);
  appendToResults(`\nPASS: version=${tc.expectedNewVersion}, bump=${tc.expectedBumpType}, Job succeeded\n`);
}

console.log(`\n${"=".repeat(60)}`);
if (allPassed) {
  console.log("ALL TEST CASES PASSED");
  appendToResults("\n" + "=".repeat(60) + "\nALL TEST CASES PASSED\n");
} else {
  console.error("SOME TEST CASES FAILED — see act-result.txt for details");
  appendToResults("\n" + "=".repeat(60) + "\nSOME TEST CASES FAILED\n");
  process.exit(1);
}

File: results/2026-05-06_173435/tasks/11-semantic-version-bumper/default-sonnet46-1m-medium/generated-code/run_act_tests.py
"""
Act test harness for the semantic version bumper.

For each test case:
  1. Create a temp git repo with the project files + test-specific fixture data
  2. Run `act push --rm` inside it
  3. Assert exit code 0, "Job succeeded", and the exact expected NEW_VERSION value
  4. Append delimited output to act-result.txt in the CWD

Usage:
  python3 run_act_tests.py
"""
import os
import re
import shutil
import subprocess
import sys
import tempfile
from pathlib import Path

WORKSPACE = Path(__file__).parent.resolve()
ACT_RESULT_FILE = WORKSPACE / "act-result.txt"

# Files from the project that must be present in every test repo
PROJECT_FILES = [
    "bump_version.py",
    "tests/__init__.py",
    "tests/test_bump_version.py",
    "tests/test_workflow.py",
    ".github/workflows/semantic-version-bumper.yml",
    ".actrc",
    "fixtures/commits_patch.txt",
    "fixtures/commits_minor.txt",
    "fixtures/commits_major.txt",
    "fixtures/commits_mixed.txt",
]

TEST_CASES = [
    {
        "name": "patch-bump",
        "description": "fix commits only -> patch bump",
        "version": "1.0.0",
        "commits": "fix: correct typo in README\nfix(auth): handle null pointer\n",
        "expected_version": "1.0.1",
    },
    {
        "name": "minor-bump",
        "description": "feat commit -> minor bump",
        "version": "1.1.0",
        "commits": "feat: add user profile API\nfix: correct date formatting\n",
        "expected_version": "1.2.0",
    },
    {
        "name": "major-bump",
        "description": "breaking commit -> major bump",
        "version": "2.0.0",
        "commits": "feat!: redesign public REST API\nfix: update internal logger\n",
        "expected_version": "3.0.0",
    },
    {
        "name": "minor-beats-patch",
        "description": "mixed fix+feat commits -> minor bump (feat wins)",
        "version": "0.5.3",
        "commits": "fix: memory leak\nfeat: dark mode\nfix: typo\n",
        "expected_version": "0.6.0",
    },
]


def copy_project_to(dest: Path) -> None:
    """Copy required project files into dest, preserving directory structure."""
    for rel in PROJECT_FILES:
        src = WORKSPACE / rel
        if not src.exists():
            print(f"  WARNING: source file missing: {src}")
            continue
        dst = dest / rel
        dst.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(src, dst)
    # Copy .actrc so act uses the correct container image
    actrc_src = WORKSPACE / ".actrc"
    if actrc_src.exists():
        shutil.copy2(actrc_src, dest / ".actrc")


def setup_test_repo(tmp_dir: Path, version: str, commits: str) -> None:
    """Initialise a git repo in tmp_dir with test-specific VERSION and commits.txt."""
    copy_project_to(tmp_dir)
    (tmp_dir / "VERSION").write_text(version + "\n")
    (tmp_dir / "commits.txt").write_text(commits)

    subprocess.run(["git", "init"], cwd=tmp_dir, check=True, capture_output=True)
    subprocess.run(["git", "config", "user.email", "test@example.com"],
                   cwd=tmp_dir, check=True, capture_output=True)
    subprocess.run(["git", "config", "user.name", "Test"],
                   cwd=tmp_dir, check=True, capture_output=True)
    subprocess.run(["git", "add", "-A"], cwd=tmp_dir, check=True, capture_output=True)
    subprocess.run(["git", "commit", "-m", "test fixture commit"],
                   cwd=tmp_dir, check=True, capture_output=True)


def run_act(tmp_dir: Path) -> tuple[int, str]:
    """Run `act push --rm` in tmp_dir and return (exit_code, output)."""
    result = subprocess.run(
        ["act", "push", "--rm", "--pull=false"],
        cwd=tmp_dir,
        capture_output=True,
        text=True,
        timeout=300,
    )
    combined = result.stdout + result.stderr
    return result.returncode, combined


def assert_test_case(case: dict, exit_code: int, output: str) -> list[str]:
    """Return list of assertion failures (empty = all passed)."""
    failures = []

    if exit_code != 0:
        failures.append(f"act exited with code {exit_code} (expected 0)")

    if "Job succeeded" not in output:
        failures.append("'Job succeeded' not found in act output")

    expected = case["expected_version"]
    if f"NEW_VERSION={expected}" not in output:
        failures.append(
            f"Expected 'NEW_VERSION={expected}' in output but not found"
        )

    return failures


def main() -> int:
    print("=== Act Test Harness: Semantic Version Bumper ===\n")

    # Truncate/create act-result.txt
    ACT_RESULT_FILE.write_text("")

    all_passed = True

    for case in TEST_CASES:
        name = case["name"]
        print(f"--- Test case: {name} ({case['description']}) ---")

        with tempfile.TemporaryDirectory() as tmp:
            tmp_path = Path(tmp)
            print(f"  Setting up repo in {tmp_path}")
            try:
                setup_test_repo(tmp_path, case["version"], case["commits"])
            except subprocess.CalledProcessError as exc:
                print(f"  ERROR: git setup failed: {exc}")
                all_passed = False
                continue

            print(f"  Running: act push --rm")
            try:
                exit_code, output = run_act(tmp_path)
            except subprocess.TimeoutExpired:
                print("  ERROR: act timed out after 300s")
                all_passed = False
                with ACT_RESULT_FILE.open("a") as f:
                    f.write(f"\n{'='*60}\nTEST CASE: {name}\nRESULT: TIMEOUT\n{'='*60}\n")
                continue

            # Append to act-result.txt
            with ACT_RESULT_FILE.open("a") as f:
                f.write(f"\n{'='*60}\n")
                f.write(f"TEST CASE: {name}\n")
                f.write(f"DESCRIPTION: {case['description']}\n")
                f.write(f"INPUT VERSION: {case['version']}\n")
                f.write(f"EXPECTED VERSION: {case['expected_version']}\n")
                f.write(f"ACT EXIT CODE: {exit_code}\n")
                f.write(f"{'='*60}\n")
                f.write(output)
                f.write(f"\n{'='*60}\n")

            failures = assert_test_case(case, exit_code, output)

            if failures:
                all_passed = False
                print(f"  FAILED:")
                for fail in failures:
                    print(f"    - {fail}")
                # Print last 30 lines of output for debugging
                lines = output.splitlines()
                print("  Last 30 lines of act output:")
                for line in lines[-30:]:
                    print(f"    {line}")
            else:
                print(f"  PASSED: NEW_VERSION={case['expected_version']}, Job succeeded")

        print()

    print("=== Summary ===")
    if all_passed:
        print("ALL TEST CASES PASSED")
        return 0
    else:
        print("SOME TEST CASES FAILED")
        return 1


if __name__ == "__main__":
    sys.exit(main())
